application/ledger_query_service/src/ethereum/queries.rs

```diff
@@ -18,80 +19,36 @@ use futures::{
 use std::sync::Arc;
 
 #[derive(Serialize, Deserialize, Clone, Default, Debug)]
-pub struct EthereumTransactionQuery {
+pub struct TransactionQuery {
     from_address: Option<Address>,
     to_address: Option<Address>,
     is_contract_creation: Option<bool>,
     transaction_data: Option<Bytes>,
     transaction_data_length: Option<usize>,
 }
 
-impl QueryType for EthereumTransactionQuery {
-    fn route() -> &'static str {
-        "transactions"
-    }
-}
-
-impl ShouldExpand for EthereumTransactionQuery {
-    fn should_expand(params: &QueryParams) -> bool {
-        params.expand_results
-    }
-}
+type Topics = Vec<H256>;
 
-fn clean_0x(s: &str) -> &str {
-    if s.starts_with("0x") {
-        &s[2..]
-    } else {
-        s
-    }
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct ReceiptQuery {
+    topics: Vec<Topics>,
```

Comment:
If you want to filter for the redemption of our ERC20 HTLC and to ensure that the tokens have been transferred this should look like this: 
`topics: [[sha256(Redeemed)], [sha256(Transferred)]]`
You would want these to be in two separate logs but in the same transaction. 

Note: this will evolve in #563

Reply:
> You would want these to be in two separate logs but in the same transaction.
I got confused by this because this doesn't seem to be what the code below does.
I read https://codeburst.io/deep-dive-into-ethereum-logs-a8d2047c7371 which has cleared up quite a few things on this topic.

From what I understood, this is actually what we want:

```rust
struct EventMatcher {
    contract_address: Option<Address>,
    topics: Vec<Option<Topic>>
}
struct ReceiptQuery {
      event_matchers : Vec<EventMatcher>
}
```

Then when you match a particular log you do something like:

```rust
if let Some(address) = event_matcher.address {
      if log.address != address {
            return false
      }
}
return event_matcher.topics.iter().enumerate().all(|i, topic| {
      match topic {
            Some(topic)  =>  log.topics[i] == topic,
             None => true
      }
})
```